hal/src/nRF52840/ble_hal.cpp

```diff
@@ -15,739 +15,3147 @@
+    std::unique_ptr<Observer> observer;                     /**< BLE Observer instance. */
+    std::unique_ptr<ConnectionsManager> connectionsMgr;     /**< BLE connections manager instance. */
+    std::unique_ptr<GattServer> gatts;                      /**< BLE GATT server instance. */
+    std::unique_ptr<GattClient> gattc;                      /**< BLE GATT client instance. */
 
-// Maximum acceptable connection interval in 1.25 ms units
-const auto MAX_CONN_INTERVAL = MSEC_TO_UNITS(45, UNIT_1_25_MS);
+    static BleObject& getInstance(void);
+    int init(void);
+    int selectAntenna(hal_ble_ant_type_t antenna) const;
 
-// Connection supervisory timeout in 10 ms units
-const auto CONN_SUP_TIMEOUT = MSEC_TO_UNITS(5000, UNIT_10_MS);
+private:
+    static bool initialized_;
+
+    BleObject();
+    ~BleObject() = default;
+    static int toPlatformUUID(const hal_ble_uuid_t* halUuid, ble_uuid_t* uuid);
+    static int toHalUUID(const ble_uuid_t* uuid, hal_ble_uuid_t* halUuid);
+};
 
-// Slave latency (a number of LL connection events)
-const auto SLAVE_LATENCY = 0;
+class BleObject::BleEventDispatcher {
+public:
+    typedef void (*BleSpecificEventHandler)(void* data, void* context);
 
-// Maximum number of attempts to agree on the connection interval
-const auto MAX_CONN_PARAM_UPDATE_ATTEMPTS = 2;
+    class EventMessage {
+    public:
+        hal_ble_evts_t evt;                 /**< BLE event data. */
+        BleSpecificEventHandler handler;    /**< BLE specific event handler. */
+        void* context;                      /**< BLE specific event context. */
+        on_ble_evt_cb_t hook;               /**< Internal hook function after event being dispatched. */
+        void* hookContext;                  /**< Internal hook function context. */
+        EventMessage() {
+            handler = nullptr;
+            context = nullptr;
+            hook = nullptr;
+            hookContext = nullptr;
+        }
+        ~EventMessage() {}
+    };
 
-// Application's BLE observer priority
-const auto BLE_OBSERVER_PRIO = 3;
+    BleEventDispatcher(BleObject* obj);
+    ~BleEventDispatcher() = default;
+    int init(void);
+    int enqueue(EventMessage& msg);
+    void onGenericEventCallback(on_ble_evt_cb_t cb, void* context);
 
-// An application-specific tag identifying the SoftDevice BLE configuration (can't be 0)
-const auto CONN_CFG_TAG = 1;
+private:
+    class BleGenericEventHandler {
+    public:
+        on_ble_evt_cb_t handler;
+        void* context;
+        BleGenericEventHandler() {
+            handler = nullptr;
+            context = nullptr;
+        }
+        ~BleGenericEventHandler() = default;
+    };
 
-// Default MTU size
-const auto DEFAULT_MTU = 23;
+    os_queue_t evtQueue_;                                   /**< BLE event queue. */
+    os_thread_t evtThread_;                                 /**< BLE event thread. */
+    BleObject* bleObj_;                                     /**< Pointer to the BLE Object. */
```

Comment:
This applies to all the classes that have this member variable. What's the point of storing the pointer if `BleObject` anyways is a singleton and at any point in time can be obtained by calling `BleObject::instance()` ?

Reply:
👍